File: sources/frontend/src/models/Resources.ts
export interface Resources {
  cpu: number;
  memory: number;
  storage: number;
}

export function resourcesString({ cpu, memory, storage }: Resources) {
  return `${cpu} | ${memory} GB | ${storage} GB`;
}

export const ZeroResources = {
  cpu: 0,
  memory: 0,
  storage: 0,
};

export function minus(op1: Resources, op2: Resources) {
  return { cpu: op1.cpu - op2.cpu, memory: op1.memory - op2.memory, storage: op1.storage - op2.storage };
}

File: sources/backend/Expenses/Expenses.Service/Domain/ValueObjects/Resources.cs
﻿using AcademyCloud.Shared;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace AcademyCloud.Expenses.Domain.ValueObjects
{
    public class Resources: ValueObject
    {
        /// <summary>
        /// CPU Core.
        /// </summary>
        public int Cpu { get; set; }
        
        /// <summary>
        /// Memory in GB.
        /// </summary>
        public int Memory { get; set; }

        /// <summary>
        /// Storage in GB
        /// </summary>
        public int Storage { get; set; }

        protected override IEnumerable<object> GetAtomicValues()
        {
            yield return Cpu;
            yield return Memory;
            yield return Storage;
        }

        public Resources(int cpu, int memory, int storage)
        {
            Cpu = cpu;
            Memory = memory;
            Storage = storage;
        }

        public static Resources operator +(Resources op1, Resources op2)
        {
            return new Resources(op1.Cpu + op2.Cpu, op1.Memory + op2.Memory, op1.Storage + op2.Storage);
        }

        public static Resources Zero => new Resources(0, 0, 0);
    }
}
